hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CompactingMemStore.java

```diff
@@ -0,0 +1,565 @@
+            + FlushLargeStoresPolicy.HREGION_COLUMNFAMILY_FLUSH_SIZE_LOWER_BOUND + " for table "
+            + getRegionServices().getTableDesc().getTableName() + ":" + flushedSizeLowerBoundString + ". "
+            + nfe + ", use global config(" + flushSizeLowerBound + ") instead");
+      }
+    }
+  }
+
+  public static long getSegmentSize(Segment segment) {
+    return segment.getSize() - DEEP_OVERHEAD_PER_PIPELINE_ITEM;
+  }
+
+  public static long getSegmentListSize(LinkedList<? extends Segment> list) {
+    long res = 0;
+    for (Segment segment : list) {
+      res += getSegmentSize(segment);
+    }
+    return res;
+  }
+
+  /**
+   * @return Total memory occupied by this MemStore.
+   * This is not thread safe and the memstore may be changed while computing its size.
+   * It is the responsibility of the caller to make sure this doesn't happen.
+   */
+  @Override
+  public long size() {
+    long res = 0;
+    for (Segment item : getListOfSegments()) {
+      res += item.getSize();
+    }
+    return res;
+  }
+
+  /**
+   * This method is called when it is clear that the flush to disk is completed.
+   * The store may do any post-flush actions at this point.
+   * One example is to update the wal with sequence number that is known only at the store level.
+   */
+  @Override public void finalizeFlush() {
+    updateLowestUnflushedSequenceIdInWal(false);
+  }
+
+  @Override public boolean isCompactingMemStore() {
+    return true;
+  }
+
+  /**
+   * Push the current active memstore segment into the pipeline
+   * and create a snapshot of the tail of current compaction pipeline
+   * Snapshot must be cleared by call to {@link #clearSnapshot}.
+   * {@link #clearSnapshot(long)}.
+   * @param flushOpSeqId the sequence id that is attached to the flush operation in the wal
+   *
+   * @return {@link MemStoreSnapshot}
+   */
+  @Override
+  public MemStoreSnapshot snapshot(long flushOpSeqId) {
+    MutableSegment active = getActive();
+    // If snapshot currently has entries, then flusher failed or didn't call
+    // cleanup.  Log a warning.
+    if (!getSnapshot().isEmpty()) {
+      LOG.warn("Snapshot called again without clearing previous. " +
+          "Doing nothing. Another ongoing flush or did we fail last attempt?");
+    } else {
+      LOG.info("FLUSHING TO DISK: region "+ getRegionServices().getRegionInfo().getRegionNameAsString()
+          + "store: "+ Bytes.toString(getFamilyName()));
+      stopCompact();
+      pushActiveToPipeline(active);
+      this.snapshotId = EnvironmentEdgeManager.currentTime();
+      pushTailToSnapshot();
+    }
+    return new MemStoreSnapshot(this.snapshotId, getSnapshot());
+  }
+
+  /**
+   * On flush, how much memory we will clear.
+   * @return size of data that is going to be flushed
+   */
+  @Override public long getFlushableSize() {
+    long snapshotSize = getSnapshot().getSize();
+    if(snapshotSize == 0) {
+      snapshotSize = pipeline.getTailSize();
```

Comment:
Is it OK that the actual size would be different, when snapshot() is invoked, due to compaction being finished by then ... ?